File: tests/go/kms_test.go
package gcplocaltest

import (
	"encoding/base64"
	"encoding/json"
	"net/http"
	"testing"

	"github.com/GuitarWag/gcp-local/tests/go/testutil"
)

func TestKMSEncryptDecryptRoundTrip(t *testing.T) {
	em := testutil.Start(t)
	base := "http://" + em.Host + "/v1/projects/" + project + "/locations/us-central1"

	resp, body := doJSON(t, http.MethodPost, base+"/keyRings?keyRingId=ring1", nil)
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("create keyring: %d %s", resp.StatusCode, body)
	}

	resp, body = doJSON(t, http.MethodPost, base+"/keyRings/ring1/cryptoKeys?cryptoKeyId=key1", map[string]any{
		"purpose": "ENCRYPT_DECRYPT",
	})
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("create key: %d %s", resp.StatusCode, body)
	}

	plain := []byte("hello kms")
	resp, body = doJSON(t, http.MethodPost, base+"/keyRings/ring1/cryptoKeys/key1:encrypt", map[string]any{
		"plaintext": base64.StdEncoding.EncodeToString(plain),
	})
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("encrypt: %d %s", resp.StatusCode, body)
	}
	var er struct {
		Ciphertext string `json:"ciphertext"`
	}
	if err := json.Unmarshal(body, &er); err != nil {
		t.Fatalf("decode encrypt: %v", err)
	}
	if er.Ciphertext == "" {
		t.Fatal("empty ciphertext")
	}

	resp, body = doJSON(t, http.MethodPost, base+"/keyRings/ring1/cryptoKeys/key1:decrypt", map[string]any{
		"ciphertext": er.Ciphertext,
	})
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("decrypt: %d %s", resp.StatusCode, body)
	}
	var dr struct {
		Plaintext string `json:"plaintext"`
	}
	_ = json.Unmarshal(body, &dr)
	out, _ := base64.StdEncoding.DecodeString(dr.Plaintext)
	if string(out) != "hello kms" {
		t.Errorf("decrypted = %q", string(out))
	}
}

File: tests/typescript/kms.test.ts
import { describe, expect, it } from "vitest";

const PROJECT = "local-project";

function host(): string {
  const h = process.env.GCP_LOCAL_HOST;
  if (!h) throw new Error("GCP_LOCAL_HOST not set");
  return h;
}

async function j(method: string, url: string, body?: unknown) {
  const res = await fetch(url, {
    method,
    headers: { "Content-Type": "application/json" },
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  const text = await res.text();
  return { status: res.status, body: text ? JSON.parse(text) : undefined };
}

describe("kms", () => {
  const base = () => `http://${host()}/v1/projects/${PROJECT}/locations/us/keyRings`;

  it("encrypt and decrypt round-trip", async () => {
    expect((await j("POST", `${base()}?keyRingId=ts-ring`)).status).toBe(200);
    expect((await j("POST", `${base()}/ts-ring/cryptoKeys?cryptoKeyId=k`, { purpose: "ENCRYPT_DECRYPT" })).status).toBe(200);

    const plain = Buffer.from("hello kms ts").toString("base64");
    const enc = await j("POST", `${base()}/ts-ring/cryptoKeys/k:encrypt`, { plaintext: plain });
    expect(enc.status).toBe(200);
    const ct: string = enc.body.ciphertext;
    expect(ct).toBeTruthy();

    const dec = await j("POST", `${base()}/ts-ring/cryptoKeys/k:decrypt`, { ciphertext: ct });
    expect(dec.status).toBe(200);
    expect(Buffer.from(dec.body.plaintext, "base64").toString()).toBe("hello kms ts");
  });

  it("rejects tampered ciphertext", async () => {
    expect((await j("POST", `${base()}?keyRingId=ts-tamper`)).status).toBe(200);
    expect((await j("POST", `${base()}/ts-tamper/cryptoKeys?cryptoKeyId=k`, {})).status).toBe(200);
    const enc = await j("POST", `${base()}/ts-tamper/cryptoKeys/k:encrypt`, {
      plaintext: Buffer.from("x").toString("base64"),
    });
    const raw = Buffer.from(enc.body.ciphertext, "base64");
    raw[raw.length - 1] ^= 0xff;
    const dec = await j("POST", `${base()}/ts-tamper/cryptoKeys/k:decrypt`, {
      ciphertext: raw.toString("base64"),
    });
    expect(dec.status).not.toBe(200);
  });
});
